src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_SUBSET_ITERATORS_H
+#define SHOGUN_SUBSET_ITERATORS_H
+
+namespace shogun
+{
+	class CLabels;
+	class CFeatures;
+
+	template <
+	    typename T, std::enable_if<
+	                    (std::is_base_of<T, CFeatures>::value) ||
+	                    (std::is_base_of<T, CLabels>::value)>* = nullptr>
+	class SubsetIterator
+	{
+	public:
+		SubsetIterator(T* pointer, index_t idx = 0)
+		{
+			if (pointer->get_subset_stack()->get_last_subset() != nullptr)
+				m_subset = pointer->get_subset_stack()
+				               ->get_last_subset()
+				               ->get_subset_idx();
+			m_idx = idx;
+			// need to keep this variable so that the vector destructor isn't
+			// called
+			m_argsorted_subset = CMath::argsort(m_subset);
+			m_argsorted_subset_iter = m_argsorted_subset.begin();
+			m_ptr = pointer;
+//			SG_REF(m_ptr)
```

Comment:
@karlnapf not sure about this referencing because what happens if the feature is on the stack?

Reply:
ideally this would actually use the actual feature matrix no?

Reply:
Objects never are on stack so should be ok